fix tage

jensen-yan opened the pull request:
1. Fixed tage phr folding history calculation bug, there will be a problem when histLen is greater than folding history, it should only be calculated with hash of histLen length, not full hash, so as to pass history check, no performance impact.
2. Add new DebugFlags=TAGEHistory
3. Fix unit tests Adapt new PHR algorithm, run through all unit tests
4. Simplify tage two-way replacement algorithm, delete LFSR, directly query the useful bits of the two ways on the spot, and find one that is 0 and you can allocate it. Remove LFSR performance improvement of 0.03 points.
5. Add use_alt_on_na, performance improves by 0.06 points.

jensen-yan committed:
cpu-o3: pred: fix phr fold history

- use histLen to get the real hash, reference to
FoldedHist::check -> fold -> use histLen real history.

Change-Id: I2e33c39e6086eaa92fbae625ec56858f9d95bd49

jensen-yan committed:
cpu-o3: add TAGEHistory debug flag and update logging

- Introduced a new debug flag 'TAGEHistory' for enhanced debugging of TAGE history updates.
- Updated logging statements to utilize the new debug flag for better traceability in history updates and checks.
- Ensured that debugFlagOn is set to true for improved visibility during debugging sessions.

Change-Id: I5b3428cce35e07a6b78b31c86f08b3c5e3f94d6f

jensen-yan committed:
cpu-o3: BTBTAGE: simplify 2-way allocation; remove mask

  - Remove AllocationResult and generateAllocationMask; stop table-level
    gating by per-way useful masks.
  - New set-associative policy (no LFSR, no LRU):
      1. allocate to invalid way;
      2. else allocate to (useful==0 && weak counter);
      3. else apply one-step age penalty to (useful==0 && strong), no alloc.
  - Miss considers all ways; hit starts from main_table+1 and still scans
    all ways at that set (no restriction to hitWay).
  - Keep useful-mask recording for global reset logic only; not used for
    allocation decisions.
  - Stats: do not count updateAllocFailure when there is no table to try
    (start_table >= numPredictors); only updateAllocFailureNoValidTable.

Change-Id: I5fb73f0ae9eaca296fdc86ab982b6762e475ff70

jensen-yan committed:
cpu-o3:  fix tage unit test

- Introduced applyPathHistoryTaken function to streamline history updates based on branch predictions.
- Updated prediction cycle to utilize new history management logic, ensuring accurate history recovery and speculative updates.
- Improved allocation behavior with multiple ways, ensuring no replacement occurs without eligible slots.

Change-Id: I8cc611127553dd12b3558be69d62779487f413d5

jensen-yan committed:
cpu-o3: Implement tage use_alt_on_na gating

- Add per-branch use_alt_on_na table (128 entries, 7-bit signed counters)
  to gate alt prediction when provider is weak (counter == 0 or -1).
- Index the table with branch PC; threshold is >= 0 → choose alt.
  Initialize counters to 0.
- Train only when provider is weak: increment if alt is correct,
  decrement otherwise.
- Simplify statistics: remove updateUseAltOnNa{,Inc,Dec} declarations
  and registrations; keep updateProviderNa, updateUseAltOnNaUpdated,
  and updateUseAltOnNa{Correct,Wrong}.
- Change useAlt to std::vector<short> and add constants
  useAltOnNaSize/useAltOnNaWidth.
Rationale: adaptively select alt vs. main when providers are weak, improving
accuracy during cold/aliasing phases with minimal overhead.

Change-Id: I2af8271dd591a509d28ca7120ae3defa843c5be9

jensen-yan commented:
@codex review

eastonman commented on src/cpu/pred/btb/btb_tage.hh:
```diff
@@ -250,8 +244,11 @@ class BTBTAGE : public TimedBaseBTBPredictor
-    std::vector<std::vector<short>> useAlt;
+    // Table for tracking when to use alternative prediction on provider weak
+    // use_alt_on_na: indexed by PC, 7-bit signed saturating counter [-64, 63]
+    static constexpr unsigned useAltOnNaSize = 128;
+    static constexpr unsigned useAltOnNaWidth = 7;
```
Why not add to Python param?

eastonman commented on src/cpu/pred/btb/btb_tage.hh:
```diff
@@ -260,7 +257,7 @@ class BTBTAGE : public TimedBaseBTBPredictor
 
     // Debug flag
-    bool debugFlagOn{false};
+    bool debugFlagOn{true};
```
Is this needed?

jensen-yan replied:
Required, open by default, and then use debug flags such as TAGEHisotry to control printf

eastonman replied:
Why not just remove this flag and use gem5's debug flags only.

jensen-yan replied:
ok, i will fix it

jensen-yan committed:
cpu-o3: use python params useAltnNA[skip ci]

- Added `useAltOnNaSize` and `useAltOnNaWidth` parameters to the BTBTAGE class for improved alternative prediction handling.
- Deleted debugFlagOn flags.

Change-Id: I808807f888284ea937ed9ef260fa48c5e0e8e783
